### PR TITLE
AirfRANS: PCGrad gradient surgery for surface/vol conflict

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AirfRANS training jointly minimizes two objectives: `surface_mse` (boundary layer prediction) and `vol_mse` (volume field prediction). These gradients may point in conflicting directions — when their cosine similarity is negative, naive summation partially cancels both signals, slowing convergence and potentially preventing convergence to Pareto-optimal solutions.

**PCGrad** (Yu et al., 2020, "Gradient Surgery for Multi-Task Learning") addresses this directly: for each task pair where gradients conflict (cos-sim < 0), it projects the conflicting gradient onto the normal plane of the other, nulling out the conflicting component before accumulation. This is the most well-validated method for multi-task gradient conflict in the literature.

Expected impact: cleaner gradient signal → faster convergence and better surface/vol trade-off. The surface constraint (≤ 0.000296) is hard; PCGrad should help us push vol_mse below the 0.0017 external target while holding surface.

Paper: https://arxiv.org/abs/2001.06782

## Instructions

Implement PCGrad for AirfRANS training. The key change is splitting the joint loss into two separate backward passes and projecting conflicting gradients before the optimizer step.

**PCGrad implementation sketch:**

```python
def pcgrad_step(model, optimizer, surface_loss, vol_loss):
    # Compute gradients for each task separately
    optimizer.zero_grad()
    surface_loss.backward(retain_graph=True)
    grads_surface = [p.grad.clone() if p.grad is not None else None for p in model.parameters()]
    
    optimizer.zero_grad()
    vol_loss.backward()
    grads_vol = [p.grad.clone() if p.grad is not None else None for p in model.parameters()]
    
    # Project conflicting gradients
    projected_surface = []
    projected_vol = []
    for g_s, g_v in zip(grads_surface, grads_vol):
        if g_s is None or g_v is None:
            projected_surface.append(g_s)
            projected_vol.append(g_v)
            continue
        
        # Project g_s onto normal plane of g_v if conflicting
        cos_sim = (g_s * g_v).sum() / (g_v.norm() * g_s.norm() + 1e-8)
        if cos_sim < 0:
            g_s = g_s - (g_s * g_v).sum() / (g_v.norm()**2 + 1e-8) * g_v
        
        # Project g_v onto normal plane of g_s if conflicting
        cos_sim2 = (g_v * g_s).sum() / (g_s.norm() * g_v.norm() + 1e-8)
        if cos_sim2 < 0:
            g_v = g_v - (g_v * g_s).sum() / (g_s.norm()**2 + 1e-8) * g_s
        
        projected_surface.append(g_s)
        projected_vol.append(g_v)
    
    # Accumulate projected gradients
    optimizer.zero_grad()
    for p, g_s, g_v in zip(model.parameters(), projected_surface, projected_vol):
        if g_s is not None and g_v is not None:
            p.grad = g_s + g_v
        elif g_s is not None:
            p.grad = g_s
        elif g_v is not None:
            p.grad = g_v
    
    # Clip and step
    torch.nn.utils.clip_grad_norm_(model.parameters(), grad_clip)
    optimizer.step()
```

**Trial 1 — PCGrad with champion config (vol-weight=10x):**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 \
  --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-loss-weight 10.0 \
  --use-pcgrad \
  --wandb_group af-pcgrad
```

**Trial 2 — PCGrad + vol-weight=15x (push harder on volume):**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 \
  --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-loss-weight 15.0 \
  --use-pcgrad \
  --wandb_group af-pcgrad
```

**Smoke test:** Run 5 epochs first to verify PCGrad is actually being applied (log the fraction of conflicting gradient pairs per step — if it's near 0, PCGrad had no effect and something is wrong).

Add a `--use-pcgrad` flag to enable/disable PCGrad without touching the rest of the training loop. The flag should default to False so the baseline is unaffected.

**Important:** The two backward passes will roughly double memory usage per step. If OOM, reduce batch size accordingly while keeping total gradient accumulation steps equivalent.

## Baseline

Current AirfRANS champion (PR #3135, W&B run: `sh2zyfwr`):
- `val_primary/surface_mse`: **0.000296** (hard constraint — must not exceed this)
- `val_primary/vol_mse`: **0.002039** (target: < 0.0017 to match external benchmark)

Config: 2L/256d/4H, AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, vol-weight=10x, Fourier

Reproduce baseline:
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 \
  --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-loss-weight 10.0
```

W&B baseline: https://wandb.ai/senpai-research/icml2026/runs/sh2zyfwr

**Success criteria:** surface_mse ≤ 0.000296 AND vol_mse < 0.002039 (beat baseline), ideally vol_mse < 0.0017 (match external target).

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT jin:

## Results

### Trial 1: PCGrad + vol-weight=10x

| Metric | Baseline | PCGrad vol10x (ep 158) | Delta |
|--------|----------|----------------------|-------|
| `val_primary/surface_mse` | **0.000296** | 0.001926 | +550% (6.5x worse) |
| `full_val/volume_mse` | **0.002039** | 0.012078 | +492% (5.9x worse) |

- **W&B run:** `p89omv2u` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/p89omv2u))
- **PCGrad conflict rate:** 29–56% (mean ~35%), confirming genuine gradient conflict
- **Peak GPU memory:** ~25 GB (GPU 0)
- **Epochs completed:** 168+ (still running)
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 \
  --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --volume-loss-weight 10.0 \
  --use-pcgrad \
  --agent jin --wandb-name "jin/af-pcgrad-vol10x" --wandb-group "af-pcgrad"
```

### Trial 2: PCGrad + vol-weight=15x

| Metric | Baseline | PCGrad vol15x (ep 159) | Delta |
|--------|----------|----------------------|-------|
| `val_primary/surface_mse` | **0.000296** | 0.001510 | +410% (5.1x worse) |
| `full_val/volume_mse` | **0.002039** | 0.012408 | +508% (6.1x worse) |

- **W&B run:** `8cq4fd0i` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/8cq4fd0i))
- **PCGrad conflict rate:** 33–55% (mean ~39%), consistent conflict throughout training
- **Peak GPU memory:** ~55 GB (GPU 1)
- **Epochs completed:** 168+ (still running)
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 \
  --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --volume-loss-weight 15.0 \
  --use-pcgrad \
  --agent jin --wandb-name "jin/af-pcgrad-vol15x" --wandb-group "af-pcgrad"
```

## What happened — Negative result

**PCGrad did NOT beat baseline on either metric.** After 168+ epochs, both trials are 5–6x worse than the baseline champion on both surface_mse and vol_mse.

### Key findings

1. **Gradient conflict is real:** PCGrad detects 30–40% conflicting gradient pairs per step throughout training, confirming the hypothesis that surface and volume gradients genuinely conflict. This rate persists even into late training.

2. **PCGrad dramatically slows convergence.** Two compounding factors:
   - **Disabling `torch.compile`:** PCGrad requires `retain_graph=True` for two separate backward passes, which is incompatible with torch.compile's donated buffer optimization. This alone significantly slows per-epoch training speed.
   - **Reduced effective gradient magnitude:** When gradients conflict, PCGrad projects them onto the normal plane of the other task, zeroing out conflicting components. This reduces the effective gradient signal, slowing convergence.

3. **The surface/vol trade-off is not improved.** Despite resolving gradient conflicts, PCGrad does not find a better Pareto frontier. The baseline's simple weighted sum with vol-weight=10x achieves far better results, suggesting that the naive gradient mixing in this problem doesn't actually harm the joint optimization as much as the hypothesis predicted.

4. **Trial 2 (vol15x) slightly outperformed Trial 1 (vol10x)** on surface_mse (0.00151 vs 0.00193), likely because the higher volume weight creates a stronger vol gradient signal that survives the PCGrad projection better.

### Why PCGrad failed here

The core issue is that PCGrad trades convergence speed for gradient purity. For this problem, the 30-40% conflict rate means PCGrad is constantly removing gradient components, reducing the optimization signal. The baseline's naive weighted sum may produce imperfect gradient directions, but the full gradient magnitude allows it to converge much faster and to better optima within the training budget.

Additionally, the requirement to disable `torch.compile` is a significant practical disadvantage — it roughly halves training throughput.

### Bug fix included

Fixed a pre-existing `UnboundLocalError` bug where `primary_metric_key` (a function) was shadowed by a local string variable assignment later in the same function scope, causing Python to treat the function call as a reference to the not-yet-assigned local variable. Renamed the string variable to `primary_metric_str`.

## Suggested follow-ups

1. **CAGrad** (Conflict-Averse Gradient Descent, Liu et al. 2021) could be a better multi-task gradient method — it minimizes worst-case task loss improvement rather than simply nulling conflicting components. More stable when gradient magnitudes differ (as they do here with the 10x vol weight).

2. **Loss-balanced training** with adaptive per-task weights (e.g., GradNorm, uncertainty weighting) might be a better approach than gradient surgery — adjusting loss weights dynamically rather than modifying gradients post-hoc.

3. If the goal is specifically to push vol_mse below 0.0017 while holding surface constraint, a simpler approach like **increasing vol-weight beyond 15x** without PCGrad overhead might be more effective, since the baseline already proves good convergence at 10x.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3212 because ADVISOR: Closing — both PCGrad trials are 5-6x worse than baseline on both metrics (surface=0.001926/0.001510 vs 0.000296, vol=0.011/0.012 vs 0.002039). Root cause: PCGrad requires retain_graph=True which disables torch.compile, causing ~40% throughput loss. Additionally, 30-40% gradient conflict rate provides persistent magnitude reduction without improving the surface/vol Pareto frontier. The gradient conflict between surface and volume is not the actual bottleneck. Good diagnostic data — PCGrad added to AF dead ends. Bug fix for primary_metric_key shadowing noted. For next assignment, we'll focus on simpler volume closure approaches..

**Comment by tcapelle** [comment]
STUDENT jin:

## Updated Results — Final (runs completed)

Both trials finished. The final numbers are better than the earlier snapshot but still clearly a negative result.

### Final metrics (best checkpoint, test set)

| Metric | Baseline | Trial 1 vol10x | Trial 2 vol15x |
|--------|----------|----------------|----------------|
| `val_primary/surface_mse` | **0.000296** | 0.000801 (ep 320) | 0.000777 (ep 283) |
| `test_primary/surface_mse` | — | 0.000940 | 0.001299 |
| `full_test/volume_mse` | **0.002039** | 0.006649 | 0.007660 |

- **Trial 1:** 320 epochs to best checkpoint, val surface_mse 2.7x worse than baseline, test vol_mse 3.3x worse
- **Trial 2:** 283 epochs to best checkpoint, val surface_mse 2.6x worse than baseline, test vol_mse 3.8x worse

### Updated assessment

The runs converged much further than the earlier 168-epoch snapshot suggested (val surface_mse went from 0.0019/0.0015 → 0.0008/0.0008). But after 280-320 epochs, PCGrad still does NOT beat the baseline on either metric. The surface constraint (≤ 0.000296) is violated by ~2.6x, and vol_mse is ~3.3–3.8x above baseline.

The earlier analysis holds: PCGrad slows convergence dramatically (needing 3–6x more epochs to reach comparable metrics), largely due to disabling torch.compile and the gradient magnitude reduction from projection. **This is a negative result** — the hypothesis that gradient surgery would improve the surface/vol trade-off is not supported.
